### ISSUE TITLE
Add `@typing.override` to functions

### ISSUE THREAD

**Issue by kevinjqliu** [apache/iceberg-python#1310]
### Feature Request / Improvement

For example, [Catalog functions](https://github.com/apache/iceberg-python/blob/b7942a85dfb74ce3736c5088995e7bd0b996d56b/pyiceberg/catalog/__init__.py#L333) and [FileIO functions](https://github.com/apache/iceberg-python/blob/b7942a85dfb74ce3736c5088995e7bd0b996d56b/pyiceberg/io/__init__.py#L265)

Only for Python 3.12 but that's ok since we run 3.12 in CI 
https://rednafi.com/python/typing_override/

**Comment by cosmastech** [comment]
Would you recommend something like this:

```py
from typing import TYPE_CHECKING

if TYPE_CHECKING and sys.version_info >= (3, 12):
    from typing import override
else:
    # Define a no-op decorator when TYPE_CHECKING is False
    def override(func):
        return func
```

**Comment by kevinjqliu** [comment]
@cosmastech I dont think we need to do the override. For Python versions less than 3.12, I think this will just be a no-op
https://docs.python.org/3/library/typing.html#typing.override

**Comment by ayushjariyal** [comment]
Hi @kevinjqliu , is this issue still open? I’d be happy to contribute by working on it.

**Comment by nssalian** [comment]
@kevinjqliu I can take a stab at this, do you mind assigning to me?

**Comment by vishnuprakaz** [comment]
Hi, I'd like to pick this up. 

The previous PR #1312 stalled on Python 3.10/3.11 compatibility since typing.override only exists in 3.12+. 

I plan to use from typing_extensions import override  same pattern already used in typedef.py for Self which should work across all supported versions.
